File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/ClusterGroups.java
package org.apache.ignite.snippets;

import org.apache.ignite.Ignite;
import org.apache.ignite.IgniteCluster;
import org.apache.ignite.IgniteCompute;
import org.apache.ignite.Ignition;
import org.apache.ignite.cluster.ClusterGroup;

public class ClusterGroups {

    void remoteNodes() {
        // tag::remote-nodes[]
        Ignite ignite = Ignition.ignite();

        IgniteCluster cluster = ignite.cluster();

        // Get compute instance which will only execute
        // over remote nodes, i.e. all the nodes except for this one.
        IgniteCompute compute = ignite.compute(cluster.forRemotes());

        // Broadcast to all remote nodes and print the ID of the node
        // on which this closure is executing.
        compute.broadcast(() -> System.out.println("Hello Node: " + ignite.cluster().localNode().id()));
        // end::remote-nodes[]
    }

    void example(Ignite ignite) {
        // tag::examples[]
        IgniteCluster cluster = ignite.cluster();

        // All nodes on which the cache with name "myCache" is deployed,
        // either in client or server mode.
        ClusterGroup cacheGroup = cluster.forCacheNodes("myCache");

        // All data nodes responsible for caching data for "myCache".
        ClusterGroup dataGroup = cluster.forDataNodes("myCache");

        // All client nodes that can access "myCache".
        ClusterGroup clientGroup = cluster.forClientNodes("myCache");

        // end::examples[]
    }
}

File: docs/_docs/code-snippets/dotnet/ClusterGroups.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using Apache.Ignite.Core;
using Apache.Ignite.Core.Compute;
using Apache.Ignite.Core.Discovery.Tcp;
using Apache.Ignite.Core.Discovery.Tcp.Static;

namespace dotnet_helloworld
{
    public class ClusterGroups
    {
        // tag::broadcastAction[]
        class PrintNodeIdAction : IComputeAction
        {
            public void Invoke()
            {
                Console.WriteLine("Hello node: " +
                                  Ignition.GetIgnite().GetCluster().GetLocalNode().Id);
            }
        }

        public static void RemotesBroadcastDemo()
        {
            var ignite = Ignition.Start();

            var cluster = ignite.GetCluster();

            // Get compute instance which will only execute
            // over remote nodes, i.e. all the nodes except for this one.
            var compute = cluster.ForRemotes().GetCompute();

            // Broadcast to all remote nodes and print the ID of the node
            // on which this closure is executing.
            compute.Broadcast(new PrintNodeIdAction());
        }
        // end::broadcastAction[]

        public static void ClusterGroupsDemo()
        {
            var ignite = Ignition.Start(
                new IgniteConfiguration
                {
                    DiscoverySpi = new TcpDiscoverySpi
                    {
                        LocalPort = 48500,
                        LocalPortRange = 20,
                        IpFinder = new TcpDiscoveryStaticIpFinder
                        {
                            Endpoints = new[]
                            {
                                "127.0.0.1:48500..48520"
                            }
                        }
                    }
                }
            );
    
            // tag::clusterGroups[]
            var cluster = ignite.GetCluster();
            
            // All nodes on which cache with name "myCache" is deployed,
            // either in client or server mode.
            var cacheGroup = cluster.ForCacheNodes("myCache");

            // All data nodes responsible for caching data for "myCache".
            var dataGroup = cluster.ForDataNodes("myCache");

            // All client nodes that access "myCache".
            var clientGroup = cluster.ForClientNodes("myCache");
            // end::clusterGroups[]
        }
    }
}
